File: src/test/java/io/github/aalsanie/codes/OutcomeDefinitionTest.java
package io.github.aalsanie.codes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class OutcomeDefinitionTest {
    @Test
    void createsCustomDefinition() {
        OutcomeDefinition definition = OutcomeDefinition.custom(
            "com.example.orders",
            "ORDER_REJECTED",
            OutcomeState.FAILED,
            "Order rejected."
        );
        assertEquals("com.example.orders:ORDER_REJECTED", definition.getCode().getValue());
        assertEquals(OutcomeState.FAILED, definition.getState());
        assertEquals("Order rejected.", definition.getDefaultMessage());
        assertTrue(definition.toString().contains("ORDER_REJECTED"));
    }

    @Test
    void customDefinitionRejectsReservedNamespaceAndBadMessages() {
        assertThrows(
            IllegalArgumentException.class,
            () -> OutcomeDefinition.custom(StandardOutcomes.NAMESPACE, "CUSTOM", OutcomeState.FAILED, "No.")
        );
        assertThrows(
            IllegalArgumentException.class,
            () -> OutcomeDefinition.custom("io.github.aalsanie.codes.child", "CUSTOM", OutcomeState.FAILED, "No.")
        );
        assertThrows(
            IllegalArgumentException.class,
            () -> OutcomeDefinition.custom("com.example", "CUSTOM", OutcomeState.FAILED, " ")
        );
        assertThrows(
            IllegalArgumentException.class,
            () -> OutcomeDefinition.custom("com.example", "CUSTOM", OutcomeState.FAILED, "x".repeat(1025))
        );
        assertThrows(
            IllegalArgumentException.class,
            () -> OutcomeDefinition.custom("com.example", "CUSTOM", OutcomeState.FAILED, "bad\0message")
        );
        assertThrows(NullPointerException.class, () -> OutcomeDefinition.custom((OutcomeCode) null, OutcomeState.FAILED, "No."));
        assertThrows(NullPointerException.class, () -> OutcomeDefinition.custom(OutcomeCode.of("com.example", "X"), null, "No."));
    }

    @Test
    void standardDefinitionIsSingleton() {
        assertSame(StandardOutcomes.NOT_FOUND, OutcomeDefinition.standard("NOT_FOUND"));
        assertThrows(IllegalStateException.class, () -> OutcomeDefinition.standard("NOPE"));
    }
}

File: src/test/kotlin/io/github/aalsanie/codes/OutcomeDefinitionTest.kt
package io.github.aalsanie.codes

import org.junit.jupiter.api.Test

class OutcomeDefinitionTest {
    @Test fun createsCustomDefinition() {
        val definition = OutcomeDefinition.custom("com.example", "DECLINED", OutcomeState.FAILED, "Declined.")
        assertEquals(OutcomeCode.of("com.example", "DECLINED"), definition.code)
        assertEquals(OutcomeState.FAILED, definition.state)
        assertEquals("Declined.", definition.defaultMessage)
    }

    @Test fun createsCustomDefinitionFromCode() {
        val code = OutcomeCode.of("com.example", "READY")
        val definition = OutcomeDefinition.custom(code, OutcomeState.SUCCEEDED, "Ready.")
        assertSame(code, definition.code)
    }

    @Test fun rejectsRootReservedNamespace() {
        assertFails<IllegalArgumentException> {
            OutcomeDefinition.custom("io.github.aalsanie.codes", "FAKE", OutcomeState.FAILED, "No.")
        }
    }

    @Test fun rejectsNestedReservedNamespace() {
        assertFails<IllegalArgumentException> {
            OutcomeDefinition.custom("io.github.aalsanie.codes.standard", "OK", OutcomeState.FAILED, "Forged.")
        }
    }

    @Test fun allowsLookalikeNonReservedNamespace() {
        val definition = OutcomeDefinition.custom("io.github.aalsanie.codesx", "OK", OutcomeState.FAILED, "Custom.")
        assertEquals("io.github.aalsanie.codesx:OK", definition.code.value)
    }

    @Test fun rejectsBlankDefaultMessage() {
        assertFails<IllegalArgumentException> {
            OutcomeDefinition.custom("com.example", "CODE", OutcomeState.FAILED, " ")
        }
    }

    @Test fun rejectsOversizedDefaultMessage() {
        assertFails<IllegalArgumentException> {
            OutcomeDefinition.custom("com.example", "CODE", OutcomeState.FAILED, "x".repeat(1025))
        }
    }

    @Test fun rejectsNulDefaultMessage() {
        assertFails<IllegalArgumentException> {
            OutcomeDefinition.custom("com.example", "CODE", OutcomeState.FAILED, "bad\u0000message")
        }
    }

    @Test fun toStringContainsStableFields() {
        val definition = OutcomeDefinition.custom("com.example", "CODE", OutcomeState.PENDING, "Pending.")
        assertTrue(definition.toString().contains("com.example:CODE"))
        assertTrue(definition.toString().contains("PENDING"))
    }
}
